docs/design/audit-v2.md

```diff
@@ -0,0 +1,93 @@
+
+## Design
+
+We will create an audit observer that uses the metainfo loop, and this observer will create a reservoir sample of segments for every node. This audit observer will replace the existing method of auditing per byte. The observer will loop through all of metainfo and build a reservoir cache for each node. The audit will then pick a configurable number of random nodes, then a random segment to audit from each of the nodes' reservoirs.
+
+If each segment generates 80 pieces on average, every time we pick a segment, we're not only auditing one specific node, we're also auditing 79 other nodes. The chance of a node appearing in a segment's pointer is proportional to the amount of data that the node actually stores. The more data that the node stores, the more chance it will be audited. We will set a minimum number of audits for unvetted nodes, and expect more audits for nodes that store more data.
+
+After selection, the rest of the auditing process will occur the same way as it does currently: picking a random segment, picking a random stripe, downloading all erasure shares associated with that stripe and using Berlekamp-Welch algorithm (via the Infectious library) to verify that they haven't been altered.
+The chances of selecting the same stripe are rare, but it's normal/expected behavior if occasionally a stripe is audited more than once.
+
+We plan to run a simulation for this new method of auditing so we can estimate appropriate settings around reservoir sampling. If we decide that we want to prefer nodes with less audits, then we will implement the power of two choices, in which we randomly select two nodes, then choose the one with less audits.
+This would still require tracking number of audits, but it would prevent having to sort and query all nodes by audit count, which could cause undesirable behavior. For example, when new nodes join the network, the audit system could become stuck auditing only new nodes, and ignoring more established nodes.
+
+We are expecting close to 3 audits per day for unvetted nodes. The satellite currently issues one audit every 30 seconds, the current interval, which gives us close to 3,000 audits per day. There are about 1,000 nodes on the network currently. This could mean that the default size of a reservoir should be three, if we assume a full iteration of reservoirs in one day.
+
+### Selection via Reservoir Sampling
+
+Reservoir sampling: a family of algorithms for randomly choosing a sample of items with uniform probability from a stream of data (of unknown size).
+
+As the audit observer uses the metainfo loop to iterate through the pointerdb, we're going through all segments and creating reservoirs per node, and filling the reservoirs with segments to audit.
+
+In order to increase the amount of audits for unvetted nodes, we can build larger reservoir samples for unvetted nodes.
+Two configurations for reservoir sampling: number of segments per unvetted nodes, and for vetted.
```

Comment:
Reading more about reservoir sampling, I think we can actually have it working with only one single slot per reservior.

The audit observer will constantly update the reservoirs as the metainfo loop iterate over the metainfo db. Due to the nature of reservior sampling, this should give an equal probability for each relevant pointer to be in the reservior at the time the audit service makes a selection.

When the audit service selects a node for auditing, it will just pick the pointer that is currently in the node's reservoir, then select a random stripe from it, etc.

Having just one slot in the reservoir will save some memory on the satellite and will simplify the implementation.

Reply:
We can configure the reservoir in the current implementation to just have a size of 1. Then we can just adjust that if we need to.

Reply:
The question should we go with a more complex implementation if a single slot reservior is sufficient?

Reply:
Do you think the implementation would be significantly simpler if we limited ourselves to a sample size of 1?
I think since most of it is already written, we would just have to adjust a config value to do a reservoir size of 1. Then if we ever decided that needs to change, it will be easy. I like the more flexible implementation.

Reply:
If we go with a single slot reservoirs then a simple `map[storj.NodeID]storj.Path` can hold all reservoirs.

Reply:
We still need to keep track of the number of elements seen for each reservoir. And the calculation of the random index would be the same as it is the other way.

Reply:
it's already not a huge amount of code so I would also be in favor of keeping the implementation adjustable to increase the reservoir size. we also still need to do the simulation thing to make more decisions about config.

Reply:
OK. You are right about this. Perhaps, it makes sense to have a generic Reservoir type and use it with size 1 for auditing.